File: LAB 6_5/Python/lab6.5_3.py
def sieve_of_eratosthenes(lst):
    max_number = max(lst)
    is_prime = [True] * (max_number + 1)
    is_prime[0] = is_prime[1] = False

    for num in range(2, int(max_number ** 0.5) + 1):
        if is_prime[num]:
            for mult in range(num * num, max_number + 1, num):
                is_prime[mult] = False

    donut = [number for number in lst if is_prime[number]]
    return donut

File: LAB 6_5/C++/lab6.5_3.cpp
void sieve_of_eratosthenes(int donut[], int n) {

    int maximum = 0;

    for (int d = 0; d < n; d++) {

        if (donut[d] > maximum) {
            maximum = donut[d];
        }
    }

    bool prime[maximum + 1];

    for (int d = 0; d <= maximum; d++) {
        prime[d] = true;
    }

    prime[0] = false;
    prime[1] = false;

    for (int p = 2; p * p <= maximum; p++) {

        if (prime[p]) {
            for (int d = p * 2; d <= maximum; d += p) {
                prime[d] = false;
            }
        }
    }

    for (int d = 0; d < n; d++) {
        if (prime[donut[d]]) {
            cout << donut[d] << " ";
        }
    }
}
